File: LeetCode/Trees and Graphs/200-number-of-islands.js
/**
 * @param {character[][]} grid
 * @return {number}
 */
var numIslands = function(grid) {
    let total = 0;
    
    let dfs = function(i, j) {
        
        if (i >= grid.length || i < 0) {
            return 0;
        }
        if (j >= grid[i].length || j < 0) {
            return 0;
        }
        if (grid[i][j] == 0) {
            return 0;
        }
        grid[i][j] = 0;
        dfs(i+1,j)
        dfs(i-1,j)
        dfs(i, j+1)
        dfs(i, j-1);
    }


    
    for (let i = 0; i < grid.length; i++) {
        for (let j = 0; j < grid[i].length; j++) {
            if (grid[i][j] == 1) {
                total++;
                dfs(i, j);
            }
        }
    }
    return total;
};

File: LeetCode/Trees and Graphs/200-number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        count = 0
        if len(grid) <= 0:
            return 0
        height = len(grid)
        width = len(grid[0])
        
        def dfs(i, j):
            if grid[i][j] == '0':
                return
            else:
                grid[i][j] = '0'
            if i + 1 < height:
                dfs(i+1,j)
            if i - 1 >= 0:
                dfs(i-1,j)
            if j+1 < width:
                dfs(i, j+1)
            if j-1 >= 0:
                dfs(i, j - 1)
        
        for i in range(len(grid)):
            for j in range(len(grid[i])):
                if grid[i][j] == '0':
                    continue
                dfs(i, j)
                count += 1
        
        
        return count
        